engine/notifier_test.go

```diff
@@ -147,30 +144,49 @@ func TestNotifier_AllWorkProcessed(t *testing.T) {
 			go func() {
 				for consumedWork.Load() < totalWork {
 					<-notifier.Channel()
+				L:
 					for {
 						select {
 						case <-pendingWorkQueue:
 							consumedWork.Inc()
 						default:
-							break
+							break L
 						}
 					}
 				}
 			}()
 		}
 
-		time.Sleep(1 * time.Millisecond)
+		time.Sleep(10 * time.Millisecond)
 		start.Done() // start routines to push work
 
 		// require that all work is eventually consumed
-		require.Eventuallyf(t,
-			func() bool { return consumedWork.Load() == totalWork },
-			3*time.Second, 100*time.Millisecond,
-			"only consumed %d units of work but expecting %d", consumedWork.Load(), totalWork,
-		)
+		if !conditionEventuallySatisfied(func() bool { return consumedWork.Load() == totalWork }, 3*time.Second, 100*time.Millisecond) {
+			require.Fail(t, "timed out", "only consumed %d units of work but expecting %d", consumedWork.Load(), totalWork)
+		}
 	}
 
 	for r := 0; r < 100; r++ {
 		t.Run(fmt.Sprintf("run %d", r), singleTestRun)
 	}
 }
+
+func conditionEventuallySatisfied(condition func() bool, waitFor time.Duration, tick time.Duration) bool {
+	done := make(chan struct{})
+
+	go func() {
+		for range time.Tick(tick) {
+			if condition() {
+				close(done)
+				return
+			}
+		}
+	}()
+
+	select {
+	case <-time.After(waitFor):
+		return false
+	case <-done:
+		return true
+	}
+}
```

Comment:
Could you explain the reasoning behind adding this method. To me, it looks like it is replicating _exactly_ the functionality of `require.Eventuallyf` (?)

Reply:
> The call to consumedWork.Load() is evaluated at the time of the function call (at which point it is still 0 since no work has been consumed yet)

In the [old version](https://github.com/onflow/flow-go/blob/master/engine/notifier_test.go#L166-L170), the call to `consumedWork.Load()` is evaluated when `require.Eventuallyf` is called. This occurs **before** most of the actual logic in the goroutines occurs, at which point the value is 0. The call to `require.Eventuallyf` only *returns* after 3 second timeout, but the `consumedWork.Load()` has already been evaluated (and in fact the string interpolation has already happened). 

In other words, the error message "only consumed 0 units of work but expecting 100" has already been created before we even enter the `require.Eventuallyf` function, since it needs to be passed as an argument to the function.

Instead, what we really want is to evaluate `consumedWork.Load()` only **after** the timeout has occured. The only way to do this is to create a function which will tell us whether we timed out waiting for the condition, and then create the error message based on the result of that function.

Reply:
Perhaps this makes it more clear:

The old code:

```
require.Eventuallyf(t,
	func() bool { return consumedWork.Load() == totalWork },
	3*time.Second, 100*time.Millisecond,
	"only consumed %d units of work but expecting %d", consumedWork.Load(), totalWork,
)
```

is actually equivalent to

```
errorMsg := fmt.Sprintf("only consumed %d units of work but expecting %d", consumedWork.Load(), totalWork)

require.Eventuallyf(t,
	func() bool { return consumedWork.Load() == totalWork },
	3*time.Second, 100*time.Millisecond,
	errorMsg
)
```

Clearly, the error message will not contain the correct value here :)

Reply:
Note, the created notifier has [1 buffer](
https://github.com/onflow/flow-go/blob/58bdf615876666b990388ef3017c1e7badc84745/engine/notifier.go#L31)

so that if all consumers are busy, we can still buffer 1 notification from the producer, so that when the consumer becomes available, it will process the buffered notification. 

Therefore, in your example:
```
Producer 3 pushes to queue and sends notification (dropped)
...
Consumer 1 finishes reading from queue and incrementing counter
Consumer 1 waiting for notification...
```

is actually:
```
Producer 3 pushes to queue and sends notification (buffered)
...
Consumer 1 finishes reading from queue and incrementing counter
Consumer 1 processes the buffered notification
```

Reply:
I'm not sure if that is how `require.Eventually` works. If we look at this source code here, it's similar to what you've implemented.

```golang
func Eventually(t TestingT, condition func() bool, waitFor time.Duration, tick time.Duration, msgAndArgs ...interface{}) bool {
	if h, ok := t.(tHelper); ok {
		h.Helper()
	}

	ch := make(chan bool, 1)

	timer := time.NewTimer(waitFor)
	defer timer.Stop()

	ticker := time.NewTicker(tick)
	defer ticker.Stop()

	for tick := ticker.C; ; {
		select {
		case <-timer.C:
			return Fail(t, "Condition never satisfied", msgAndArgs...)
		case <-tick:
			tick = nil
			go func() { ch <- condition() }()
		case v := <-ch:
			if v {
				return true
			}
			tick = ticker.C
		}
	}
}
```

Reply:
But think about what the argument `msgAndArgs` holds when the function is called. Does it hold the value of `consumedWork.Load()` from the future? This is impossible. It holds the current value of `consumedWork.Load()` (which has already executed), which is 0

Reply:
As for your first point, yes my understanding of the code was not entirely correct, and the original PR summary is no longer accurate.

Reply:
>But think about what the argument msgAndArgs holds when the function is called. Does it hold the value of consumedWork.Load() from the future? This is impossible. It holds the current value of consumedWork.Load() (which has already executed), which is 0

Yes, You are right. I fixed it in my PR https://github.com/onflow/flow-go/pull/1049/commits/974dfb4dfb9cbe90bbd6a22dda1c7f8ff9bea22d